### DISCUSSION TITLE
Persistent Indices

### DISCUSSION THREAD

**Discussion by SartorialOffense** [duckdb/duckdb#3413] [General]
Given the progress this project has made and the immeasurable help it has been to my current work, it pains me to ask about an estimated timeline for anything, but... Is there a super rough idea when the ART indices will be persistable - say in 2022 or 2023? I have "hybrid" queries where range scans of one 250M row table is followed by a join to another 250M table where an index lookup is dramatically faster in my testing.  But I can't pay the penalty of rebuilding the index for every load given the size. Or if there is a way to force the optimizer to pick a range type scan for the join rather than just a hash? Because of the structure/ordering of the data I suspect this would be faster especially since I am usually only loading in ~0.01% of the rows in the second table per query.

I can live with it for now, but in the future I need to scale beyond 250M rows and it would be super helpful to know if it was a 2022 thing or beyond.

Thanks!

**Comment by pdet** [comment]
Yes, you are right, we have no support for buffer managing ART-Indexes.

We had some discussions today about it, I think it should be added in the next release.

The basic battle plan is to write Blocks with the ```MetaBlock Writer``` in a depth-first fashion.

e.g.,
```
                            [Node 4(c)]
             [Node 4 (b)]                 [Leaf (d)]
[Leaf (a)]
```
would be serialized as:
```block 0
[Leaf (a)|  Node 4 (b) Off:[0] | Leaf (d) | Node 4 (c) Off:[1,2]]
```

In the index metadata, we also must point to the block_id and offset of the root node.

With this type of serialization, we can lazily load the ART-Index by checking if a child exists, if it exists and is not yet in memory we then use the buffer manager to bring it to memory. Since the ART-index is pretty protected with mutexes, doing this in different threads shouldn't be problematic.

The only question here is how to later update these blocks properly, I guess that initially, we can just always serialize the full tree during the checkpoint.

**Reply by SartorialOffense** [reply]
Thank you very much for the update, very exciting. 

> I guess that initially, we can just always serialize the full tree during the checkpoint.

I think this would support a reasonable number of use cases including my own because in most instances the table is "write once".

**Reply by danz68** [reply]
There is an art implementation which is quite interesting from the persistent perspective. I am not sure if their IO logic is similar with duckdb storage mechanism to reach the same level of performance, though.

https://github.com/pmem/pmdk/tree/master/src/examples/libpmemobj/libart
